### ISSUE TITLE
Re-evaluate the design of extents and customizable buffers

### ISSUE THREAD

**Issue by sam20908** [sam20908/mpp#370]
Since we allowed customization of buffers, the extents system became practically obsolete.

- Remove different types of dynamic buffers
- Remove `mpp::type` as it's no longer necessary
- Remove allocator (that's the buffer's responsibility instead)
- Remove unnecessary `mat_base` and `mat_dyn_base` class
- Remove `get_allocator` member function and add `buffer` function to return the internal buffer instead.
- Remove constructors for `std::array` as they're not necessary anymore due to no specializations.
- Revamp tests

#### TL;DR
- Remove concept of allocators with buffers instead (buffers can take in custom allocators)
- Remove concept of extents since buffers can be customized, which means extents is deprecated (fixed matrices means the buffer is supplied with `std::array` as the buffer for example).

This is going to be a wild journey...
